File: branches/Modeller_gwt_Branch/src/org/openremote/modeler/domain/Activity.java
/* OpenRemote, the Home of the Digital Home.
* Copyright 2008-2009, OpenRemote Inc.
*
* See the contributors.txt file in the distribution for a
* full listing of individual contributors.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Affero General Public License as
* published by the Free Software Foundation, either version 3 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*/
package org.openremote.modeler.domain;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.Transient;


/**
 * The Class Activity.
 * @author allen.wei
 */
@SuppressWarnings("serial")
public class Activity extends BusinessEntity {
   
   
   /** The default name index. */
   private static int defaultNameIndex = 1;
   
   /** The name of the activity. It is the  display name on the device */
   private String name;
   
   /** The screens. */
   private List<Screen> screens = new ArrayList<Screen>();
   
   /**
    * Gets the name.
    * 
    * @return the name
    */
   public String getName() {
      return name;
   }
   
   /**
    * Sets the name.
    * 
    * @param name the new name
    */
   public void setName(String name) {
      this.name = name;
   }
   
   /**
    * Gets the screens.
    * 
    * @return the screens
    */
   public List<Screen> getScreens() {
      return screens;
   }
   
   /**
    * Sets the screens.
    * 
    * @param screens the new screens
    */
   public void setScreens(List<Screen> screens) {
      this.screens = screens;
   }
   
   /**
    * Adds the screen.
    * 
    * @param screen the screen
    */
   public void addScreen(Screen screen) {
      screens.add(screen);
   }
   
   /**
    * Delete screen.
    * 
    * @param screen the screen
    */
   public void deleteScreen(Screen screen) {
      screens.remove(screen);
   }
   
   
   /**
    * {@inheritDoc}
    * @see org.openremote.modeler.domain.BusinessEntity#getDisplayName()
    */
   @Transient
   public String getDisplayName() {
      return name;
   }
   

   /**
    * Gets the new default name when you want a new name. such as activity1.
    * 
    * @return the new default name
    */
   @Transient
   public static String getNewDefaultName() {
      return "activity" + defaultNameIndex++;
   }
}

File: tags/project/iPhone_Console/iPhone_Console_1_0_0_Beta5/Classes/Activity.m
/* OpenRemote, the Home of the Digital Home.
 * Copyright 2008-2009, OpenRemote Inc.
 * 
 * See the contributors.txt file in the distribution for a
 * full listing of individual contributors.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3.0 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */


#import "Activity.h"
#import "Screen.h"

@implementation Activity

@synthesize activityId,name, icon, screens;

#pragma mark Initializers

/**
 * Initialize according to the XML parser.
 */
- (id)initWithXMLParser:(NSXMLParser *)parser elementName:(NSString *)elementName attributes:(NSDictionary *)attributeDict parentDelegate:(NSObject *)parent {
	if (self == [super init]) {
		activityId = [[attributeDict objectForKey:@"id"] intValue];
		name = [[attributeDict objectForKey:@"name"] copy];
		icon = [[attributeDict objectForKey:@"icon"] copy]; 
		
		screens = [[NSMutableArray alloc] init];
		
		xmlParserParentDelegate = [parent retain];
		[parser setDelegate:self];
	}
	return self;
}

- (void)dealloc {
	[name release];
	[icon release];
	[screens release];
	
	[super dealloc];
}

#pragma mark Delegate methods of NSXMLParser

/**
 * Parse the screen elements by creating Screen instances for them.
 */
- (void)parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qualifiedName attributes:(NSDictionary *)attributeDict{
	if ([elementName isEqualToString:@"screen"]) {
		NSLog(@"start parse Screen");
		Screen *screen = [[Screen alloc] initWithXMLParser:parser elementName:elementName attributes:attributeDict parentDelegate:self];
		[screens addObject:screen];
		[screen release];
	}
}

/**
 * When we find an activity end element, restore the original XML parser delegate.
 */
- (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName {
	if ([elementName isEqualToString:@"activity"]) {
		NSLog(@"end parse Activity");
 		[parser setDelegate:xmlParserParentDelegate];
		[xmlParserParentDelegate release];
		xmlParserParentDelegate = nil;
	}
}

@end
